src/main/java/uk/ac/ebi/ampt2d/accessioning/commons/accessioning/BasicAccessioningServiceSaveDelegate.java

```diff
@@ -0,0 +1,58 @@
+package uk.ac.ebi.ampt2d.accessioning.commons.accessioning;
+
+import org.springframework.dao.DataIntegrityViolationException;
+import uk.ac.ebi.ampt2d.accessioning.commons.generators.ModelHashAccession;
+import uk.ac.ebi.ampt2d.accessioning.commons.persistence.DatabaseService;
+
+import java.util.HashMap;
+import java.util.List;
+import java.util.Stack;
+
+public class BasicAccessioningServiceSaveDelegate<MODEL, HASH, ACCESSION> {
+
+    private class Partition {
+
+        private int start;
+        private int end;
+
+        public Partition(int start, int end) {
+            this.start = start;
+            this.end = end;
+        }
+    }
+
+    private final DatabaseService<MODEL, HASH, ACCESSION> dbService;
+
+    public BasicAccessioningServiceSaveDelegate(DatabaseService<MODEL, HASH, ACCESSION> dbService) {
+        this.dbService = dbService;
+    }
+
+    public SaveResponse<ACCESSION, MODEL> doSaveAccessions(
+            List<ModelHashAccession<MODEL, HASH, ACCESSION>> modelHashAccessions) {
+        Stack<Partition> partitions = new Stack<>();
+        partitions.add(new Partition(0, modelHashAccessions.size()));
+        HashMap<ACCESSION, MODEL> savedAccessions = new HashMap<ACCESSION, MODEL>();
+        HashMap<ACCESSION, MODEL> notSavedAccessions = new HashMap<ACCESSION, MODEL>();
+
+        while (!partitions.isEmpty()) {
+            Partition partition = partitions.pop();
+            final List<ModelHashAccession<MODEL, HASH, ACCESSION>> partitionToSave = modelHashAccessions
+                    .subList(partition.start, partition.end);
+            try {
+                dbService.save(partitionToSave);
+                partitionToSave.stream().forEach(mha -> savedAccessions.put(mha.accession(), mha.model()));
```

Comment:
so you make a stack with only one element: a "partition" that is a range from 0 to n where n is how many elements we have to store, and you copy the list modelHashAccessions into partitionsToSave and then you save the list?

I see in the catch that you do this to split the data in smaller ranges until it succeedes, but why would it fail if the region is too big?

Reply:
It is an iterative version of a binary partition. The problem appears when an object has been already accessioned, same hash message, same elements, but different id. In hashed accessions it won't happen but in monotonic accessioning can happen when multiple instances of the application.

The idea is if the save works, all elements have been stored, otherwise, because it is a transactional operation all saves are dropped, so we need to retry but remove the element that provocked the error, but JPA doesn't tell you what element was responsible, so, you need to split the batch of elements, then try to save the first part, then the second, and split them again until they fail or has failed and there is only one element on the partition. In that case the element is stored as, this element can't be stored. It is a recursive algorithm just made in iterate form.